File: src/main/java/org/example/binarysearch/RotatedArrSearch.java
package org.example.binarysearch;

import java.util.Arrays;

public class RotatedArrSearch {
    public int search(int[] nums, int target) {
        if (nums.length == 1) {
            if (nums[0] == target) return 0; else return -1;
        }

        int left = 0, right = nums.length-1;

        while (left <= right) {
            int mid = left + (right - left)/2;
            if (nums[mid] == target) return mid;

            // Check to see if target is in left sorted array
            if (nums[left] <= nums[mid]) {
                if (target > nums[mid] || target < nums[left]) {
                    left = mid + 1;
                } else {
                    right = mid - 1;
                }
            } else {
               // we are in right sorted portion
                if (target < nums[mid] || target > nums[right]) {
                    right = mid - 1;
                } else {
                    left = mid + 1;
                }
            }
        }

        return -1;
    }
}

File: src/main/java/org/example/binarysearch/RotatedArrSearch.py
from typing import List


class Solution:
    def search(self, nums: List[int], target: int) -> int:
        if len(nums) == 1: return 1 if target == nums[0] else -1
        l, r = 0, len(nums) - 1

        while l <= r:
            m = (l + r) // 2
            if target == nums[m]: return m

            if nums[l] <= nums[m]:
                if target < nums[l] or target > nums[m]:
                    l = m + 1
                else:
                    r = m - 1
            else:
                if target > nums[r] or target < nums[m]:
                    r = m - 1
                else:
                    l = m + 1
        return -1


obj = Solution()
nums = [4,5,6,7,0,1,2]
target = 0
#target = 3
#nums = [1]
#target = 0
#nums = [5,1,3]
#target = 3
print(obj.search(nums, target))